tests/mbt/model/ccv_pss.qnt

```diff
@@ -0,0 +1,157 @@
+            (acc, validator) =>
+                pure val curValSet = acc._1
+                pure val accPower = acc._2
+                pure val shouldAddSamePow = acc._3
+                pure val lastPow = acc._4
+
+                pure val validatorPower = origValSet.get(validator)
+                if (validatorPower == lastPow and shouldAddSamePow) {
+                    // we should add the validator because it has the same power as the previous one,
+                    // and we add regardless of total power because we need to include all
+                    // vals with the same power if we include one of them
+                    pure val newAccPower = accPower + validatorPower
+                    (curValSet.union(Set(validator)), newAccPower, true, validatorPower)
+                } else if (validatorPower > 0 and accPower < topNPower) {
+                    // if we don't have enough power yet, add the validator to the set
+                    pure val newAccPower = accPower + validatorPower
+                    (curValSet.union(Set(validator)), newAccPower, true, validatorPower)
+                } else {
+                    // if we have enough power and we also are done adding
+                    // all validators with the same power as the lowest top N validator,
+                    // don't add them
+                    (curValSet, accPower, false, 0)
+                }
+        )
+        res._1
+    }
+
+    // Opts a validator in for a consumer chain the provider.
+    // Possible before the consumer chain starts running,
+    // and will then be applied when the consumer chain starts running.
+    pure def OptIn(currentState: ProtocolState, consumer: Chain, validator: Node): Result = {
+        pure val optedInVals = currentState.providerState.optedInVals.get(consumer)
+        pure val newOptedInVals = optedInVals.union(Set(validator))
+        Ok({
+            ...currentState,
+            providerState: {
+                ...currentState.providerState,
+                optedInVals: currentState.providerState.optedInVals.put(consumer, newOptedInVals)
+            }
+        })
+    }
+
+    // Returns true if the given validator is in the top N for the given consumer chain,
+    // and false otherwise.
+    pure def IsTopN(currentState: ProtocolState, validator: Node, consumer: Chain): bool = {
+        val proviValSet = currentState.providerState.chainState.currentValidatorSet
+        val N = currentState.providerState.topNByConsumer.get(consumer)
+
+        val topNValSet = GetTopNVals(proviValSet, N)
+
+        topNValSet.contains(validator)
+    }
+
+    // Returns true if the given validator has opted in to the given consumer chain,
+    pure def IsOptedIn(currentState: ProtocolState, validator: Node, consumer: Chain): bool = {
+        currentState.providerState.optedInVals.getOrElse(consumer, Set()).contains(validator)
+    }
+
+    // Opts a validator out. Safe to call before the consumer chain even runs.
```

Comment:
> Safe to call before the consumer chain even runs.

Regarding the implementation of this (although we do not have the code yet), I was thinking that we should not allow validators to opt out before the chain has started running. This is because, we assume validator that voted `Yes` on a `ConsumerAdditionProposal` are opted in by default, so it would be weird to give them the right to opt out before the chain starts. What do you think?

Reply:
Good point. My thought was to use this, at the time the consumer starts, to determine who voted yes on the proposal (and is thus opted in), but maybe I am cutting too many corners here and it is not as clear.
But essentially I think this is an implementation detail and maybe abstracting away like I did is ok, but I would love to know if you disagree

Reply:
> at the time the consumer starts, to determine who voted yes on the proposal (and is thus opted in) [...]
Thanks for your answer. This makes sense. We can leave as is :)